elasticsearch5/search-index/src/main/java/org/vertexium/elasticsearch5/bulk/BulkUpdateService.java

```diff
@@ -216,21 +226,99 @@ private void complete(BulkItem bulkItem, Exception exception) {
-                builder.add((UpdateRequest) actionRequest);
-            } else if (actionRequest instanceof DeleteRequest) {
-                builder.add((DeleteRequest) actionRequest);
-            } else {
-                throw new VertexiumException("unhandled request type: " + actionRequest.getClass().getName());
+
+        Map<String, List<BulkItem>> byIndexName = bulkItems.stream().collect(Collectors.groupingBy(BulkItem::getIndexName));
+        for (Map.Entry<String, List<BulkItem>> byIndexNameEntry : byIndexName.entrySet()) {
+            String indexName = byIndexNameEntry.getKey();
+            Map<String, List<BulkItem>> byType = byIndexNameEntry.getValue().stream().collect(Collectors.groupingBy(BulkItem::getType));
+
+            for (Map.Entry<String, List<BulkItem>> byTypeEntry : byType.entrySet()) {
+                String type = byTypeEntry.getKey();
+                Map<String, List<BulkItem>> byDocumentId = byTypeEntry.getValue().stream().collect(Collectors.groupingBy(BulkItem::getDocumentId));
+
+                for (Map.Entry<String, List<BulkItem>> byDocumentIdEntry : byDocumentId.entrySet()) {
+                    String documentId = byDocumentIdEntry.getKey();
+                    List<BulkItem> requestBulkItems = byDocumentIdEntry.getValue();
+                    Map<Class<? extends BulkItem>, List<BulkItem>> byBulkItemType = requestBulkItems.stream().collect(Collectors.groupingBy(BulkItem::getClass));
+                    if (byBulkItemType.size() > 1) {
+                        throw new VertexiumException(
+                            "Cannot mix bulk item types in a batch: "
+                                + byBulkItemType.keySet().stream().map(Class::getName).collect(Collectors.joining(","))
+                        );
+                    }
+                    for (Map.Entry<Class<? extends BulkItem>, List<BulkItem>> byBulkItemTypeEntry : byBulkItemType.entrySet()) {
+                        Class<? extends BulkItem> bulkItemType = byBulkItemTypeEntry.getKey();
+                        if (bulkItemType == UpdateBulkItem.class) {
+                            Map<String, String> source = new HashMap<>();
+                            Map<String, Object> fieldsToSet = new HashMap<>();
+                            List<String> fieldsToRemove = new ArrayList<>();
+                            Map<String, String> fieldsToRename = new HashMap<>();
+                            List<String> additionalVisibilities = new ArrayList<>();
+                            List<String> additionalVisibilitiesToDelete = new ArrayList<>();
+
+                            boolean updateOnly = true;
+                            for (BulkItem bulkItem : requestBulkItems) {
+                                UpdateBulkItem updateBulkItem = (UpdateBulkItem) bulkItem;
+                                source.putAll(updateBulkItem.getSource());
+                                fieldsToSet.putAll(updateBulkItem.getFieldsToSet());
```

Comment:
I think you need to get a little fancier merging here. You may need to check the type of the `value`. For example, if three `BulkItem` instances want to set `first_name`:
1. `"Joe"`
1. `["Joseph"]`
1. `["J."]`
It would really need to end up as `["Joe", "Joseph","J."]`, but I think in this case it will be just `["J."]`

Reply:
In the above example, I think the SearchIndex class that build the BulkItem ensures that non-arrays aren't possible as in the first item. I'm not 100% sure though.